pkg/vere/disk.c

```diff
@@ -867,66 +868,522 @@ u3_disk_init(c3_c* pax_c, u3_disk_cb cb_u)
+   *     a. creates epoch directory
+   *     b. creates epoch version file
+   *     c. creates binary version file
+   *     d. initializes database
+   *     e. reads metadata from old database
+   *     f. writes metadata to new database
+   *     g. loads new epoch directory and sets it in log_u
+   *  2. create hard links to data.mdb and lock.mdb in 0i0/
+   *  3. rollover to new epoch (second call to u3_disk_epoc_init()) *
+   *     a. same as 1a-g but also copies current snapshot between c/d steps
+   *  4. delete backup snapshot (c3_unlink() and c3_rmdir() calls) *
+   *  5. delete old data.mdb and lock.mdb files (c3_unlink() calls) *
+   */
+
+  //  check if data.mdb is readable in log directory
+  c3_o dat_o = c3n;
+  c3_c dat_c[8193];
+  snprintf(dat_c, sizeof(dat_c), "%s/data.mdb", log_u->com_u->pax_c);
+  if ( 0 == access(dat_c, R_OK) ) {
+    dat_o = c3y;
+  }
+
+  //  check if lock.mdb is readable in log directory
+  c3_o lok_o = c3n;
+  c3_c lok_c[8193];
+  snprintf(lok_c, sizeof(dat_c), "%s/lock.mdb", log_u->com_u->pax_c);
+  if ( 0 == access(dat_c, R_OK) ) {
+    lok_o = c3y;
+  }
+
+  //  if fresh boot, initialize disk v1
+  if ( c3y == u3_Host.ops_u.nuu ) {
+    fprintf(stderr, "disk: initializing disk with v%d format\r\n", U3D_VER1);
+
+    //  initialize first epoch "0i0"
+    if ( c3n == u3_disk_epoc_init(log_u) ) {
+      fprintf(stderr, "disk: failed to initialize first epoch\r\n");
+      return c3n;
+    }
+
+    return c3y;
+  }
+  
+  //  if data.mdb is not readable, skip migration
+  if ( c3n == dat_o ) return c3y;
+
+  //  migrate existing pier which has either:
+  //  - not started the migration, or
+  //  - crashed before completing the migration
+
+  //  initialize pre-migrated lmdb
+  {
+    if ( 0 == (log_u->mdb_u = u3_lmdb_init(log_u->com_u->pax_c)) ) {
+      fprintf(stderr, "disk: failed to initialize database\r\n");
+      return c3n;
+    }
+  }
+
+  //  get first/last event numbers from pre-migrated lmdb
+  c3_d fir_d, las_d;
+  if ( c3n == u3_lmdb_gulf(log_u->mdb_u, &fir_d, &las_d) ) {
+    fprintf(stderr, "disk: failed to get first/last event numbers\r\n");
+    return c3n;
+  }
+
+  // ensure there's a current snapshot
+  if ( u3_Host.eve_d != las_d ) {
+    fprintf(stderr, "disk: migrate: error: snapshot is out of date, please "
+                    "start/shutdown your pier gracefully first\r\n");
+    fprintf(stderr, "disk: migrate: eve_d (%" PRIc3_d ") != las_d (%" PRIc3_d ")\r\n",
+                    u3_Host.eve_d, las_d);
+    return c3n;
+  }
+
+  //  initialize first epoch "0i0"
+  if ( c3n == u3_disk_epoc_init(log_u) ) {
+    fprintf(stderr, "disk: migrate: failed to initialize first epoch\r\n");
+    return c3n;
+  }
+
+  //  create hard links to data.mdb and lock.mdb in 0i0/
+  c3_c dut_c[8193], luk_c[8193];  //  old paths
+  snprintf(dut_c, sizeof(dut_c), "%s/data.mdb", log_u->com_u->pax_c);
+  snprintf(luk_c, sizeof(luk_c), "%s/lock.mdb", log_u->com_u->pax_c);
```

Comment:
You actually have unnecessary code duplication here. In the function-level scope at the top, you declare `dat_c` and `lok_c` which contain the same strings as `dut_c` and `luk_c` here. So you should just declare dut and luk at the top and use them here

Reply:
Uh, so slight change, because it looks like you collapsed the condition

Reply:
Below, you snprintf the new dat and lock locations into `dat_c` and `luk_c` already declared. _Instead_, I would just at the top of `u3_disk_migrate`:
```C
c3_c dut_c[8193], luk_c[8193];  //  old paths
snprintf(dut_c, sizeof(dut_c), "%s/data.mdb", log_u->com_u->pax_c);
snprintf(luk_c, sizeof(luk_c), "%s/lock.mdb", log_u->com_u->pax_c);

// do the stuff with access(dut_c), access(luk_c) etc

// now here, declare dat_c and lok_c
c3_c epo_c[8193], dat_c[8193], lok_c[8193];  //  new paths
snprintf(epo_c, sizeof(epo_c), "%s/0i0", log_u->com_u->pax_c);
snprintf(dat_c, sizeof(dat_c), "%s/data.mdb", epo_c);
snprintf(lok_c, sizeof(lok_c), "%s/lock.mdb", epo_c);

// ...
```

Reply:
i.e. at the beginning of this function, dat_c and lok_c store the _old_ paths and later store the new paths when dut_c and luk_c store the old paths. Just always store the old paths in dut_c and luk_c. Swapping them is confusing/unnecessary